src/crystal/system/win32/file_descriptor.cr

```diff
@@ -274,6 +273,47 @@ module Crystal::System::FileDescriptor
     end
     ret
   end
+
+  private def console_mode?(handle)
+    LibC.GetConsoleMode(handle, out _) != 0
+  end
+
+  private def read_console(handle : LibC::HANDLE, buffer : UInt16*, count : Int) : Int32
+    if 0 == LibC.ReadConsoleW(handle, buffer, count, out units_read, nil)
+      raise IO::Error.from_winerror("ReadConsoleW")
+    end
+    units_read.to_i32!
+  end
+
+  private def read_console(handle : LibC::HANDLE, slice : Bytes) : Int32
+    # Characters between U+0000 and U+FFFF have 1 UTF-16 code unit and up to 3 UTF-8 code unit.
+    # Characters between U+10000 and U+10FFFF have 2 UTF-16 code units and 4 UTF-8 code units.
+    # The following conclusions can therefore be drawn:
+    # 1. N UTF-16 code units correspond to no more than 3*N UTF-8 code units.
+    # 2. We can re-use the last 2/3 of the UTF-8 buffer to receive the UTF-16 data without worrying about data overlap.
+
+    third = slice.size // 3
+    bytes_buffer = slice.to_unsafe
+    units_buffer = (bytes_buffer + third).unsafe_as(Pointer(UInt16))
+
+    # One character may have two UTF-16 code units.
+    raise IO::Error.new("Buffer size too small") if third < 2
+
+    # Reads in two batches to guarantee that the last character is intact.
+    units_read = read_console(handle, units_buffer, third - 1)
+    return 0 if units_read < 1
+    if units_buffer[units_read - 1] & 0xFC00 == 0xD800
```

Comment:
I reckon this should have the additional condition that the entire buffer was filled (`units_read == third - 1`).
Otherweise, if there's still space in the buffer the entire surrogate pair could've been written into it. Failing to do so is weird and unexpected. Reading again probably wouldn't fix that.